### PR TITLE
Fix TimeDistributed BatchNormalization

### PR DESCRIPTION
**Pull request by waleedka**
This PR fixes issue #7466. The root cause of the problem is that TimeDistributed reshapes the inputs before passing them to BatchNormalization, and batch normalization has update ops that are conditioned on the inputs. Since the inputs to TimeDistributed are different from those passed to BatchNormalization, the conditional updates don't get executed and the mean and variance don't get updated.

Wrappers could modify the inputs in many different ways, so I implemented a generic solution on the Wrapper class to map wrapper inputs to inner layer inputs. This solution should fix the issue with BatchNormaliztion and also with any other wrapped layer that might have update ops conditioned on its inputs.

### CODE REVIEW

**Comment by QuantumLiu** [comment]
Thanks for your job.But it got 1 failing check in `TEST_MODE=PEP8`?

**Comment by waleedka** [comment]
Thanks for the heads up! I had a line over indented. It's fixed, and I also fixed another indentation issue that was already in the original code.

**Comment by fchollet** [comment]
- Please add unit test (a test that would have been failing with the previous implementation but would be passing now, as well as enough tests to have full coverage of the new code)
- Don't break lines with `\`

**Comment by milani** [comment]
Is this the same as BatchNormalization's `mode=2` in keras 1.x?

**Comment by ahundt** [comment]
is there any chance https://github.com/fchollet/keras/pull/7033 might be fixed by this?

**Comment by waleedka** [comment]
@fchollet Done! Added the unit tests and removed the \

**Comment by waleedka** [comment]
@fchollet Done! Please review.
